lumigator/backend/backend/services/datasets.py

```diff
@@ -224,17 +221,32 @@ def get_dataset(self, dataset_id: UUID) -> DatasetResponse | None:
 
         return DatasetResponse.model_validate(record)
 
-    def get_dataset_by_job_id(self, job_id: UUID) -> DatasetResponse | None:
+    def get_dataset_by_job_id(self, job_id: UUID) -> DatasetResponse:
+        """Gets the dataset associated to a job.
+
+        :param job_id: job ID to find the associated dataset for
+        :return: information on the associated dataset, for example filename, size, if it contains
+                 ground truth data.
+        :rtype: DatasetResponse
+        :raises DatasetNotAvailableError: if there is no dataset associated with the supplied job ID
+        """
         record = self._get_dataset_record_by_job_id(job_id)
         if record is None:
-            return None
+            raise DatasetNotAvailableError(f"not found using job ID: {job_id}") from None
```

Comment:
Why not the same DatasetNotFoundError?

Reply:
It was to indicate the semantics, with `DatasetNotFoundError` the dataset resource, as identified by it's ID isn't available, but here the job ID is being used to try and find a dataset with a corresponding run ID. 

It felt better to suggest that the dataset (for this job/run ID) wasn't available, to say: the job/run hasn't generated a dataset that has the attribute.